osu.Game/Rulesets/UI/GameplaySampleTriggerSource.cs

```diff
@@ -65,64 +69,61 @@ public virtual void Play()
-            // This lookup can be skipped if the last entry is still valid (in the future and not yet hit).
-            if (fallbackObject == null || fallbackObject.Result?.HasResult == true)
+            if (mostValidObject == null || isAlreadyHit(mostValidObject))
             {
                 // We need to use lifetime entries to find the next object (we can't just use `hitObjectContainer.Objects` due to pooling - it may even be empty).
                 // If required, we can make this lookup more efficient by adding support to get next-future-entry in LifetimeEntryManager.
-                fallbackObject = hitObjectContainer.Entries
-                                                   .Where(e => e.Result?.HasResult != true).MinBy(e => e.HitObject.StartTime);
-
-                if (fallbackObject != null)
-                    return getEarliestNestedObject(fallbackObject.HitObject);
+                var candidate =
+                    // Use alive entries first as an optimisation.
+                    hitObjectContainer.AliveEntries.Select(tuple => tuple.Entry).Where(e => !isAlreadyHit(e)).MinBy(e => e.HitObject.StartTime)
+                    ?? hitObjectContainer.Entries.Where(e => !isAlreadyHit(e)).MinBy(e => e.HitObject.StartTime);
 
                 // In the case there are no non-judged objects, the last hit object should be used instead.
-                fallbackObject ??= hitObjectContainer.Entries.LastOrDefault();
+                if (candidate == null)
+                {
+                    mostValidObject = hitObjectContainer.Entries.LastOrDefault();
+                }
+                else
+                {
+                    if (isCloseEnoughToCurrentTime(candidate.HitObject))
+                    {
+                        mostValidObject = candidate;
+                    }
+                    else
+                    {
+                        mostValidObject ??= hitObjectContainer.Entries.FirstOrDefault();
+                    }
+                }
             }
 
-            if (fallbackObject == null)
+            if (mostValidObject == null)
                 return null;
 
-            bool fallbackHasResult = fallbackObject.Result?.HasResult == true;
-
             // If the fallback has been judged then we want the sample from the object itself.
-            if (fallbackHasResult)
-                return fallbackObject.HitObject;
+            if (isAlreadyHit(mostValidObject))
+                return mostValidObject.HitObject;
 
-            // Else we want the earliest (including nested).
+            // Else we want the earliest valid nested.
             // In cases of nested objects, they will always have earlier sample data than their parent object.
-            return getEarliestNestedObject(fallbackObject.HitObject);
+            return getAllNested(mostValidObject.HitObject).OrderBy(h => h.StartTime).FirstOrDefault(h => h.StartTime > getReferenceTime()) ?? mostValidObject.HitObject;
```

Comment:
Looking at this again with a cool head, I'm no longer sure this is correct, and I'm no longer sure that the adjustments I made to the taiko tests are correct.

In particular I mean this test:

https://github.com/ppy/osu/blob/4ff52752082f93906df68ef0d8cb1a1048d44f92/osu.Game.Rulesets.Taiko.Tests/TestSceneDrumSampleTriggerSource.cs#L299-L318

The comment just above this line says that "nested objects will always have earlier sample data than their parent object". But there are several issues here which caused the assertion above to change in a way that contradicts that comment:

- The hitobject will always have `StartTime` earlier than its nested hitobjects. In addition `getAllNested()` does prefix recursion, which means that on any level of recursion, it will return the parent hitobject before the child hitobject, and `OrderBy()` is a stable sorting algorithm. All of this means that for all intents and purposes, what's going to be first after `OrderBy()` is the _parent hitobject itself_, and only then the nesteds.
- Even if that wasn't the case, if we're _completely before_ the object, the `FirstOrDefault()` filters everything out. So we end up with `mostValidObject.HitObject`, which again is the outermost object rather than the earliest nested one.

Notably, generally sample playback is associated with the _end time_ of an object. In light of this, I think this should read something like the following?

```suggestion
            return getAllNested(mostValidObject.HitObject).OrderBy(h => h.GetEndTime()).SkipWhile(h => h.GetEndTime() <= getReferenceTime()).FirstOrDefault();
```

- Sorting is done by end time which fixes the nested object ordering.
- `SkipWhile().FirstOrDefault()` with an inverted condition is used to skip to the bits we want (if there are any bits we want yet). If there are none, then it's a no-op, and `FirstOrDefault()` falls back to the object that ends earliest.

Reply:
Notably, you've removed the fallback to `mostValidObject.HitObject` (the parenting object). This means after the last nested tick, it won't get the main object correctly to play its sample if it hasn't been hit yet. Am I missing something here?